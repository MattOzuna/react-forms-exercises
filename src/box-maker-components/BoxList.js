import { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import "./BoxList.css";

const BoxList = () => {
  const [boxes, setBoxes] = useState([
    // { color: "red", width: 100, height: 50 },
    // { color: "blue", width: 50, height: 100 },
    // { color: "yellow", width: 100, height: 100 },
  ]);

  const AddBox = ({ color, width, height }) => {
    setBoxes((boxes) => [...boxes, { color, width, height }]);
  };

  const deleteBox = (idx) => {
    setBoxes((boxes) => boxes.toSpliced(idx, 1));
  };

  return (
    <>
      <h1>Box List</h1>
      <div className="BoxList">
        {boxes.map(({ color, width, height }, idx) => (
          <Box
            color={color}
            width={width}
            height={height}
            key={idx}
            id={idx}
            deleteBox={deleteBox}
          />
        ))}
      </div>
      <NewBoxForm AddBox={AddBox} />
    </>
  );
};

export default BoxList;
