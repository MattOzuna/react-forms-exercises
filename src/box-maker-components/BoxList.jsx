import { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import "./BoxList.css"

const BoxList = () => {
    const [boxes, setBoxes] = useState([
        {color: "red", width:100, height:50},
        {color: "blue", width:50, height:100},
        {color: "yellow", width:100, height:100},
        {color: "red", width:100, height:50},
        {color: "blue", width:50, height:100},
        {color: "yellow", width:100, height:100},
        {color: "red", width:100, height:50},
        {color: "blue", width:50, height:100},
        {color: "yellow", width:100, height:100}
    ])

    const AddBox = ({color, width, height}) => {
        setBoxes(boxes => [...boxes, {color, width, height}])
    }

    return (
        <>
            <h1>Box List</h1>
            <div className="BoxList"> 
                {boxes.map(({color, width, height}, idx) => (
                    <Box 
                        color={color}
                        width={width}
                        height={height}
                        key = {idx}
                    />
                ))}
            </div>
            <NewBoxForm 
                AddBox={AddBox}
            />
        </>
    )
}

export default BoxList