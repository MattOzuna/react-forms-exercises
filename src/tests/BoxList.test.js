import { render, fireEvent } from "@testing-library/react";
import BoxList from "../box-maker-components/BoxList";
import "@testing-library/jest-dom";

it("should render", () => {
  render(<BoxList />);
});

it("should match snapshot", () => {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("should add a box to the page", () => {
  const { getByLabelText, queryByText } = render(<BoxList />);

  const colorInput = getByLabelText("Color:");
  const WidthInput = getByLabelText("Width:");
  const HeightInput = getByLabelText("Height:");
  const formButton = queryByText("Submit");

  expect(queryByText("X")).not.toBeInTheDocument();

  fireEvent.change(colorInput, { target: { value: "White" } });
  fireEvent.change(WidthInput, { target: { value: "100" } });
  fireEvent.change(HeightInput, { target: { value: "100" } });
  fireEvent.click(formButton);

  expect(queryByText("X")).toBeInTheDocument();
});

it("should remove Box by clicking X button", () => {
  const { getByLabelText, queryByText } = render(<BoxList />);

  const colorInput = getByLabelText("Color:");
  const widthInput = getByLabelText("Width:");
  const heightInput = getByLabelText("Height:");
  const formInput = queryByText("Submit");

  fireEvent.change(colorInput, { target: { value: "White" } });
  fireEvent.change(widthInput, { target: { value: "100" } });
  fireEvent.change(heightInput, { target: { value: "100" } });
  fireEvent.click(formInput);

  expect(queryByText("X")).toBeInTheDocument();

  const removeButton = queryByText("X");
  fireEvent.click(removeButton);

  expect(queryByText("X")).not.toBeInTheDocument();
});
