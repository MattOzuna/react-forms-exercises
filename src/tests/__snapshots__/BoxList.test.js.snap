// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <h1>
    Box List
  </h1>
  <div
    class="BoxList"
  />
  <form
    class="NewBoxForm"
  >
    <label
      for="color"
    >
      Color: 
    </label>
    <input
      id="color"
      name="color"
      type="text"
      value=""
    />
    <label
      for="width"
    >
      Width: 
    </label>
    <input
      id="width"
      name="width"
      type="number"
      value=""
    />
    <label
      for="height"
    >
      Height: 
    </label>
    <input
      id="height"
      name="height"
      type="number"
      value=""
    />
    <button>
      Submit
    </button>
  </form>
</DocumentFragment>
`;
